testsuite/forge/src/backend/local/mod.rs

```diff
@@ -1,85 +1,124 @@
+                bin,
+                version,
+            }
+        })?;
+
+        versions.insert(new_version.version.clone(), new_version);
+        Ok(Self::new(versions))
     }
-}
 
-fn get_diem_node() -> Result<PathBuf> {
-    let output = Command::new("cargo")
-        .current_dir(workspace_root()?)
-        .args(&["build", "--bin=diem-node"])
-        .output()
-        .context("Failed to build diem-node")?;
-
-    if output.status.success() {
-        let bin_path = build_dir()?.join(format!("{}{}", "diem-node", env::consts::EXE_SUFFIX));
-        if !bin_path.exists() {
-            bail!(
-                "Can't find binary diem-node in expected path {:?}",
-                bin_path
-            );
-        }
+    pub fn from_revision(revision: &str) -> Result<Self> {
+        let mut versions = HashMap::new();
+        let new_version =
+            cargo::get_diem_node_binary_at_revision(revision).map(|(revision, bin)| {
+                let version = Version::new(usize::max_value(), revision.clone());
+                LocalVersion {
+                    revision,
+                    bin,
+                    version,
+                }
+            })?;
 
-        Ok(bin_path)
-    } else {
-        bail!("Faild to build diem-node");
+        versions.insert(new_version.version.clone(), new_version);
+        Ok(Self::new(versions))
     }
-}
 
-// Path to top level workspace
-fn workspace_root() -> Result<PathBuf> {
-    let mut path = build_dir()?;
-    while !path.ends_with("target") {
-        path.pop();
+    pub fn with_revision_and_workspace(revision: &str) -> Result<Self> {
+        let workspace = cargo::get_diem_node_binary_from_worktree().map(|(revision, bin)| {
+            let version = Version::new(usize::max_value(), revision.clone());
+            LocalVersion {
+                revision,
+                bin,
+                version,
+            }
+        })?;
+        let revision =
+            cargo::get_diem_node_binary_at_revision(revision).map(|(revision, bin)| {
+                let version = Version::new(usize::min_value(), revision.clone());
+                LocalVersion {
+                    revision,
+                    bin,
+                    version,
+                }
+            })?;
+
+        let mut versions = HashMap::new();
+        versions.insert(workspace.version(), workspace);
+        versions.insert(revision.version(), revision);
+        Ok(Self::new(versions))
     }
-    path.pop();
-    Ok(path)
-}
 
-// Path to the directory where build artifacts live.
-fn build_dir() -> Result<PathBuf> {
-    env::current_exe()
-        .ok()
-        .map(|mut path| {
-            path.pop();
-            if path.ends_with("deps") {
-                path.pop();
-            }
-            path
-        })
-        .context("Can't find the build directory. Cannot continue running tests")
+    /// Create a LocalFactory with a diem-node version built at the tip of upstream/main and the
```

Comment:
nit: top

Reply:
The actually correct terminology is `tip` as in the "tip of a branch", you generally wouldn't say the "top of a branch"

Reply:
wow, aha gotcha